modules/citrus-http/src/main/java/com/consol/citrus/http/message/HttpMessageConverter.java

```diff
@@ -17,158 +17,99 @@
-        }
+    @Override
+    public HttpEntity<?> convertOutbound(Message message,
+                                         HttpEndpointConfiguration endpointConfiguration,
+                                         TestContext context) {
+
+        HttpMessage httpMessage = convertOutboundMessage(message);
+
+        HttpHeaders httpHeaders = createHttpHeaders(httpMessage, endpointConfiguration);
 
         Object payload = httpMessage.getPayload();
         if (httpMessage.getStatusCode() != null) {
             return new ResponseEntity<>(payload, httpHeaders, httpMessage.getStatusCode());
         } else {
             for (Cookie cookie : httpMessage.getCookies()) {
-                httpHeaders.set("Cookie", cookie.getName() + "=" + context.replaceDynamicContentInString(cookie.getValue()));
+                httpHeaders.add(
+                        "Cookie",
+                        cookie.getName() + "=" + context.replaceDynamicContentInString(cookie.getValue()));
             }
         }
 
-        HttpMethod method = endpointConfiguration.getRequestMethod();
-        if (httpMessage.getRequestMethod() != null) {
-            method = httpMessage.getRequestMethod();
-        }
-
-        if (httpMethodSupportsBody(method)) {
-            return new HttpEntity<>(payload, httpHeaders);
-        } else {
-            return new HttpEntity<>(httpHeaders);
-        }
-    }
+        HttpMethod method = determineHttpMethod(endpointConfiguration, httpMessage);
 
-    private boolean httpMethodSupportsBody(HttpMethod method) {
-        return HttpMethod.POST.equals(method) || HttpMethod.PUT.equals(method)
-                || HttpMethod.DELETE.equals(method) || HttpMethod.PATCH.equals(method);
+        return createHttpEntity(httpHeaders, payload, method);
     }
 
     @Override
-    public HttpMessage convertInbound(HttpEntity<?> message, HttpEndpointConfiguration endpointConfiguration, TestContext context) {
+    public HttpMessage convertInbound(HttpEntity<?> message,
+                                      HttpEndpointConfiguration endpointConfiguration,
+                                      TestContext context) {
         Map<String, Object> mappedHeaders = endpointConfiguration.getHeaderMapper().toHeaders(message.getHeaders());
-        HttpMessage httpMessage = new HttpMessage(message.getBody() != null ? message.getBody() : "", convertHeaderTypes(mappedHeaders));
+        HttpMessage httpMessage = new HttpMessage(extractMessageBody(message), convertHeaderTypes(mappedHeaders));
 
         for (Map.Entry<String, String> customHeader : getCustomHeaders(message.getHeaders(), mappedHeaders).entrySet()) {
             httpMessage.setHeader(customHeader.getKey(), customHeader.getValue());
         }
 
         if (message instanceof ResponseEntity<?>) {
             httpMessage.status(((ResponseEntity<?>) message).getStatusCode());
-            httpMessage.version("HTTP/1.1"); //TODO check if we have access to version information
```

Comment:
So there's no version info? 😃

Reply:
I've added some comments to the code to make this clear. :wink: